velox/substrait/SubstraitToVeloxPlan.cpp

```diff
@@ -154,18 +206,51 @@ core::PlanNodePtr SubstraitVeloxPlanConverter::toVeloxPlan(
 
   const auto& inputType = childNode->outputType();
   int colIdx = 0;
+  // Note that Substrait projection adds the project expressions on top of the
+  // input to the projection node. Thus we need to add the input columns first
+  // and then add the projection expressions.
+
+  // First, adding the project names and expressions from the input to
+  // the project node.
+  for (uint32_t idx = 0; idx < inputType->size(); idx++) {
+    const auto& fieldName = inputType->nameOf(idx);
+    projectNames.emplace_back(fieldName);
+    expressions.emplace_back(std::make_shared<core::FieldAccessTypedExpr>(
+        inputType->childAt(idx), fieldName));
+    colIdx += 1;
```

Comment:
@vibhatha Why adding the `colIdx` here? If there are only 5 columns in our schema, this line of code will cause the `colIdx` larger than 5 at the line 227 and make the `emitProjectNames` become '[xx_5, xx_6,  xx_7,  xx_8,  xx_9]' at the line 239. And at the end, how would you re-construct the output names base on the  `emitProjectNames`?

Reply:
Even though there are 5 columns in the schema, if you had projections, the output could be more columns. Because in Substrait projections would add columns not just projected columns. Emit determines if we just want to output a selected number of columns in a given order. The reason for doing +1 operation is when we do a projection Substrait adds columns. You can see the definition [here](https://substrait.io/relations/logical_relations/#project-operation).

May I see a sample Substrait plan which would cause issues because of this change? Then I can comment more.

Reply:
@vibhatha We have a Calcite plan, which is 
```
LogicalProject(_cluster=[$0], tag1=[$2], tag2=[$3], dc=[$1], $_field=[$4])
  LogicalAggregate(group=[{0, 1, 2, 3}], $_field=[map_union_sum($4)])
    LogicalFilter(condition=[AND(=($0, 'Tesla'), SEARCH($2, Sarg['0', '1']:CHAR(1)), =($3, '0'), =($1, 'fake_sandbox'))])
      LogicalTableScan(table=[[test, _default]])
```

And the corresponding Substrait plan is 
```
extension_uris {
  extension_uri_anchor: 3
  uri: "substrait/metrics_functions.yaml"
}
extension_uris {
  extension_uri_anchor: 1
  uri: "/functions_boolean.yaml"
}
extension_uris {
  extension_uri_anchor: 2
  uri: "/functions_comparison.yaml"
}
extensions {
  extension_function {
    extension_uri_reference: 1
    name: "and:bool"
  }
}
extensions {
  extension_function {
    extension_uri_reference: 2
    function_anchor: 1
    name: "equal:any_any"
  }
}
extensions {
  extension_function {
    extension_uri_reference: 1
    function_anchor: 2
    name: "or:bool"
  }
}
extensions {
  extension_function {
    extension_uri_reference: 3
    function_anchor: 3
    name: "map_union_sum:map"
  }
}
relations {
  root {
    input {
      project {
        common {
          emit {
            output_mapping: 5
            output_mapping: 6
            output_mapping: 7
            output_mapping: 8
            output_mapping: 9
          }
        }
        input {
          aggregate {
            common {
              direct {
              }
            }
            input {
              filter {
                common {
                  direct {
                  }
                }
                input {
                  read {
                    common {
                      direct {
                      }
                    }
                    base_schema {
                      names: "_cluster"
                      names: "dc"
                      names: "tag1"
                      names: "tag2"
                      names: "$_field"
                      struct {
                        types {
                          string {
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        types {
                          string {
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        types {
                          string {
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        types {
                          string {
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        types {
                          map {
                            key {
                              i64 {
                                nullability: NULLABILITY_REQUIRED
                              }
                            }
                            value {
                              fp64 {
                                nullability: NULLABILITY_REQUIRED
                              }
                            }
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        nullability: NULLABILITY_REQUIRED
                      }
                    }
                    named_table {
                      names: "test"
                      names: "_default"
                    }
                  }
                }
                condition {
                  scalar_function {
                    output_type {
                      bool {
                        nullability: NULLABILITY_REQUIRED
                      }
                    }
                    arguments {
                      value {
                        scalar_function {
                          function_reference: 1
                          output_type {
                            bool {
                              nullability: NULLABILITY_REQUIRED
                            }
                          }
                          arguments {
                            value {
                              selection {
                                direct_reference {
                                  struct_field {
                                  }
                                }
                                root_reference {
                                }
                              }
                            }
                          }
                          arguments {
                            value {
                              cast {
                                type {
                                  string {
                                    nullability: NULLABILITY_REQUIRED
                                  }
                                }
                                input {
                                  literal {
                                    fixed_char: "Tesla"
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                    arguments {
                      value {
                        scalar_function {
                          function_reference: 2
                          output_type {
                            bool {
                              nullability: NULLABILITY_REQUIRED
                            }
                          }
                          arguments {
                            value {
                              scalar_function {
                                function_reference: 1
                                output_type {
                                  bool {
                                    nullability: NULLABILITY_REQUIRED
                                  }
                                }
                                arguments {
                                  value {
                                    selection {
                                      direct_reference {
                                        struct_field {
                                          field: 2
                                        }
                                      }
                                      root_reference {
                                      }
                                    }
                                  }
                                }
                                arguments {
                                  value {
                                    cast {
                                      type {
                                        string {
                                          nullability: NULLABILITY_REQUIRED
                                        }
                                      }
                                      input {
                                        literal {
                                          fixed_char: "0"
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                          arguments {
                            value {
                              scalar_function {
                                function_reference: 1
                                output_type {
                                  bool {
                                    nullability: NULLABILITY_REQUIRED
                                  }
                                }
                                arguments {
                                  value {
                                    selection {
                                      direct_reference {
                                        struct_field {
                                          field: 2
                                        }
                                      }
                                      root_reference {
                                      }
                                    }
                                  }
                                }
                                arguments {
                                  value {
                                    cast {
                                      type {
                                        string {
                                          nullability: NULLABILITY_REQUIRED
                                        }
                                      }
                                      input {
                                        literal {
                                          fixed_char: "1"
                                        }
                                      }
                                    }
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                    arguments {
                      value {
                        scalar_function {
                          function_reference: 1
                          output_type {
                            bool {
                              nullability: NULLABILITY_REQUIRED
                            }
                          }
                          arguments {
                            value {
                              selection {
                                direct_reference {
                                  struct_field {
                                    field: 3
                                  }
                                }
                                root_reference {
                                }
                              }
                            }
                          }
                          arguments {
                            value {
                              cast {
                                type {
                                  string {
                                    nullability: NULLABILITY_REQUIRED
                                  }
                                }
                                input {
                                  literal {
                                    fixed_char: "0"
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                    arguments {
                      value {
                        scalar_function {
                          function_reference: 1
                          output_type {
                            bool {
                              nullability: NULLABILITY_REQUIRED
                            }
                          }
                          arguments {
                            value {
                              selection {
                                direct_reference {
                                  struct_field {
                                    field: 1
                                  }
                                }
                                root_reference {
                                }
                              }
                            }
                          }
                          arguments {
                            value {
                              cast {
                                type {
                                  string {
                                    nullability: NULLABILITY_REQUIRED
                                  }
                                }
                                input {
                                  literal {
                                    fixed_char: "fake_sandbox"
                                  }
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
            groupings {
              grouping_expressions {
                selection {
                  direct_reference {
                    struct_field {
                    }
                  }
                  root_reference {
                  }
                }
              }
              grouping_expressions {
                selection {
                  direct_reference {
                    struct_field {
                      field: 1
                    }
                  }
                  root_reference {
                  }
                }
              }
              grouping_expressions {
                selection {
                  direct_reference {
                    struct_field {
                      field: 2
                    }
                  }
                  root_reference {
                  }
                }
              }
              grouping_expressions {
                selection {
                  direct_reference {
                    struct_field {
                      field: 3
                    }
                  }
                  root_reference {
                  }
                }
              }
            }
            measures {
              measure {
                function_reference: 3
                phase: AGGREGATION_PHASE_INITIAL_TO_RESULT
                output_type {
                  map {
                    key {
                      i64 {
                        nullability: NULLABILITY_REQUIRED
                      }
                    }
                    value {
                      fp64 {
                        nullability: NULLABILITY_REQUIRED
                      }
                    }
                    nullability: NULLABILITY_REQUIRED
                  }
                }
                invocation: AGGREGATION_INVOCATION_ALL
                arguments {
                  value {
                    selection {
                      direct_reference {
                        struct_field {
                          field: 4
                        }
                      }
                      root_reference {
                      }
                    }
                  }
                }
              }
            }
          }
        }
        expressions {
          selection {
            direct_reference {
              struct_field {
              }
            }
            root_reference {
            }
          }
        }
        expressions {
          selection {
            direct_reference {
              struct_field {
                field: 2
              }
            }
            root_reference {
            }
          }
        }
        expressions {
          selection {
            direct_reference {
              struct_field {
                field: 3
              }
            }
            root_reference {
            }
          }
        }
        expressions {
          selection {
            direct_reference {
              struct_field {
                field: 1
              }
            }
            root_reference {
            }
          }
        }
        expressions {
          selection {
            direct_reference {
              struct_field {
                field: 4
              }
            }
            root_reference {
            }
          }
        }
      }
    }
    names: "_cluster"
    names: "tag1"
    names: "tag2"
    names: "dc"
    names: "$_field"
  }
}
```
Base on the plan, we know that the projection names are:
```
    names: "_cluster"          // ordinal: 0
    names: "tag1"                //  ordinal: 1
    names: "tag2"               //  ordinal: 2
    names: "dc"                  //   ordinal : 3
    names: "$_field"           //   ordinal: 4
```

Before this change, the output of this plan would be like 
<img width="758" alt="image" src="https://github.com/facebookincubator/velox/assets/5697276/c1a426ac-20c6-4108-93fb-75c4fc9e5c5f">
That is, the column names are "[n3_0, n3_1, n3_2, n3_3]" and we know that `n3_0` is referenced the first projection name, which  is `_cluster`.

But after this change, the output change to this:
<img width="804" alt="image" src="https://github.com/facebookincubator/velox/assets/5697276/189f2149-8c9b-421c-81bf-e65872079fe8">
The column names change to something like "[n3_5, n3_6, n3_7, n3_8]" and we don't know which projection names they are referenced anymore.

Reply:
So looking into the calcite plan and the corresponding Substrait plan, 

```txt
LogicalProject(_cluster=[$0], tag1=[$2], tag2=[$3], dc=[$1], $_field=[$4])
  LogicalAggregate(group=[{0, 1, 2, 3}], $_field=[map_union_sum($4)])
    LogicalFilter(condition=[AND(=($0, 'Tesla'), SEARCH($2, Sarg['0', '1']:CHAR(1)), =($3, '0'), =($1, 'fake_sandbox'))])
      LogicalTableScan(table=[[test, _default]])
```

The projection is _cluster=[$0], tag1=[$2], tag2=[$3], dc=[$1], $_field=[$4], 5 projections in total. 
Aggregation introduces 4 groupby columns plus aggregation column
Filter just filters data.
Table scan loads the data from the table `test._default`. Containing the following schema. 

Schema contains 

```json
base_schema {
                      names: "_cluster"
                      names: "dc"
                      names: "tag1"
                      names: "tag2"
                      names: "$_field"
                      struct {
                        types {
                          string {
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        types {
                          string {
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        types {
                          string {
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        types {
                          string {
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        types {
                          map {
                            key {
                              i64 {
                                nullability: NULLABILITY_REQUIRED
                              }
                            }
                            value {
                              fp64 {
                                nullability: NULLABILITY_REQUIRED
                              }
                            }
                            nullability: NULLABILITY_REQUIRED
                          }
                        }
                        nullability: NULLABILITY_REQUIRED
                      }
                    }
```

There are 5 fields in the schema. 

names: "_cluster" 	0
names: "dc" 		1
names: "tag1" 	2
names: "tag2"		3
names: "$_field"	4


After Read the schema ==> `schema({("_cluster", string), ("dc", string), ("tag1", string), ("tag2", string), ("$_field", map[i64, fp64])})`
After Filter there are 5 columns (filter won't add or remove columns) ==> `schema({("_cluster", string), ("dc", string), ("tag1", string), ("tag2", string), ("$_field", map[i64, fp64])})` 
After Aggregate there are 5 columns (4 grouping columns plus aggregated columns) ==> `schema({("_cluster", string), ("dc", string), ("tag1", string), ("tag2", string), ("$map_union_sum", map[i64, fp64])})` 
Projection adds 5 columns (according to Subtrait), total columns 10 (Refer to Substrait projection definition), ==> `schema({("_cluster", string), ("dc", string), ("tag1", string), ("tag2", string), ("$map_union_sum", map[i64, fp64]), ("_cluster",string), ("tag1",string), ("tag2",string), ("dc",string), ("$_field", map[i64, fp64])})` 

Substrait always append projected fields. But your emit says you only need fields 5, 6, 7, 8, 9

From the output schema of the field we would have the indices, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, we just need 5, 6, 7, 8, 9. They are the following fields.

("_cluster",string), ("tag1",string), ("tag2",string), ("dc",string), ("$_field", map[i64, fp64])

In your case the output names should be n3_5, n3_6, n3_7, n3_9 and n3_9. The project names are clearly listed in the plan (towards the bottom most section,)

```json
    names: "_cluster"
    names: "tag1"
    names: "tag2"
    names: "dc"
    names: "$_field"
```

My curiosity is would this change your output values, do you get incorrect results (column data)?

Also, in the PR description, there is a description on the correction to project definition in Velox and adding emit feature. Hope this helps.

Reply:
@whutjs is there a way to run your program end-to-end and evaluate the differences? Just curious.

Reply:
> Projection adds 5 columns (according to Subtrait), total columns 10 (Refer to Substrait projection definition), ==> schema({("_cluster", string), ("dc", string), ("tag1", string), ("tag2", string), ("$map_union_sum", map[i64, fp64]), ("_cluster",string), ("tag1",string), ("tag2",string), ("dc",string), ("$_field", map[i64, fp64])})

I don't think the codes in `SubstraitToVeloxPlan` in velox is implemented the way you said.

Reply:
Could you please show me why it is not? This is an already merged PR, but it included test cases to cover certain angels, if we are missing something in our implementation, can you provide a reproducible test case which shows that current implementation doesn't support it properly?

Reply:
@vibhatha The important point is that the (projection) names are totally lost in velox output, including the 10 column names you said. And we cannot rebuild the names based on `[n3_5, n3_6, n3_7...]` at all. 
<img width="936" alt="image" src="https://github.com/facebookincubator/velox/assets/5697276/c7de6e96-dad9-4c5f-920e-cad716089cad">
![image](https://github.com/facebookincubator/velox/assets/5697276/55bc681a-1dfa-4caa-8e3f-7a1449dbc479)

Reply:
@vibhatha And you can use the Substrait plan I provided above to re-produce this problem. The goal is to produce an output with names `[_cluster, tag1, tag2...]`, not `[n3_5, n3_6, n3_7...]`.

Reply:
I'm not sure I understand.  Why do you need `n3_0`?  Where does this come from?  I don't see `n3_5` in the plan anywhere?  Are you looking up the name in some kind of map?

Since you have the plan's output names (an array of 5 strings), and your output has 5 columns, then the first column should be the first output name.  The second column should be the second output name.  And so on.  You shouldn't need `n3_5`.  It doesn't exist in the Substrait plan at all.

Reply:
As @westonpace mentioned the output names are already in the plan. 

They are not totally lost, the `RelRoot` has the output names accurately. It doesn't mean that the projection implementation and emit implementation are inaccurate. If you take a look at the current names, it precisely outputs the field labels as expected. Also please take a look at the PR description, this fixes an issue in the Velox projection definition along with adding emit feature. Also this PR hasn't made any changes to existing mechanisms which define output names of Velox computed plan.

Reply:
> I'm not sure I understand. Why do you need `n3_0`? Where does this come from? I don't see `n3_5` in the plan anywhere? Are you looking up the name in some kind of map?
> 
> Since you have the plan's output names (an array of 5 strings), and your output has 5 columns, then the first column should be the first output name. The second column should be the second output name. And so on. You shouldn't need `n3_5`. It doesn't exist in the Substrait plan at all.

That's the point.  `n3_5` doesn't exist in the Substrait plan at all, but it just show up in velox output instead of `_custer`.

Reply:
@whutjs Named are redefined, and please check this part, https://github.com/facebookincubator/velox/blob/main/velox/substrait/SubstraitToVeloxPlan.cpp#L438. Substrait is index-based, so to maintain the orginal names through takes extra efforts.

Reply:
It is not a part of this PR. This PR doesn't change the output names of the Velox Substrait plan execution. As I previously mentioned, this only added project definition and emit feature. I think it would be the best to create an issue.

Reply:
The Substrait plan contains accurate output names, and if Velox computation doesn't produce column names, it means, it is not using those fields to update the `RowVector` field names. And as previously mentioned, this PR didn't address that.

Reply:
@rui-mo we could extract output names from the `RelRoot` and update the output names? Would it be too complicated to handle?

Reply:
@vibhatha I think adding a simple projection on top of the Velox plan can do that. We haven't supported that because the output data is what we mainly care about.

Reply:
But I would propose, if output names are not accurately represented, wouldn't it be better to make a PR to fix it. The main doubt is as you said, will it be hard to maintain that kind of a change? Appreciate your thoughts.

Reply:
> @whutjs Named are redefined, and please check this part, https://github.com/facebookincubator/velox/blob/main/velox/substrait/SubstraitToVeloxPlan.cpp#L438. Substrait is index-based, so to maintain the orginal names through takes extra efforts.

@rui-mo Thanks to clarify. @vibhatha   In velox, `SubstraitParser::makeNodeName(int node_id, int col_idx)` return a name in the format of `n3_0`.  And by calling the `getIdxFromNodeName()` method, we can extract the columnIndex from the name.
<img width="740" alt="image" src="https://github.com/facebookincubator/velox/assets/5697276/4e766499-a8b9-40ac-9c1f-00cfc5232b6b">

So before the change, we can call `getIdxFromNodeName('n3_1')` and know that the column index of this name is `1`, and that's how we get the actual projection name as `tag1` from:
```
 names: "_cluster"
    names: "tag1"
    names: "tag2"
    names: "dc"
    names: "$_field"
``` 

But after the change, if you call `getIdxFromNodeName()` with `n3_7`, you would get a column index 7, which is out of range.
By the way, I am totally ok with the `emit ordering` feature in this PR, but I don't see the necessity of `colIdx += 1` in https://github.com/facebookincubator/velox/blob/main/velox/substrait/SubstraitToVeloxPlan.cpp#L214 .  And my original question is why adding this line of code? In my option, removing this line of code seems doesn't affect the program at all.

Reply:
No that is not a correct interpretation, that produces in correct definition for projection. If you don't have that emit and run the query, you should get 10 columns according to Substrait. I wouldn't propose changing the code to fix an incorrect and older representation of Substrait in Velox. It is better to adapt to the accurate definition. The `RelRoot` contains the names for your output, why can't that information be used? That is the definition of Substrait. I have referenced numerous links to reference definitions of Substrait. I would suggest to adopt the definition precisely. 

cc @mbasmanova

Reply:
> By the way, I am totally ok with the `emit ordering` feature in this PR, but I don't see the necessity of `colIdx += 1` in https://github.com/facebookincubator/velox/blob/main/velox/substrait/SubstraitToVeloxPlan.cpp#L214 . And my original question is why adding this line of code? 

Emit is a feature in Substrait, if a particular consumer supports Substrait it must be precisely used. That is why we made this PR.

> In my option, removing this line of code seems doesn't affect the program at all.

On the other hand, it would be better to propose a fix and see if it works for the current test cases and validate the thesis.

Reply:
Folks, it would be nice to create a GitHub issue and move the discussion there. Overall I find Substrait is difficult and confusing to use and support. Also, it is not designed to represent optimized physical plans. I suggest to consider using some other representation for these, maybe, serialized Velox plans.

Reply:
Thank you @mbasmanova .  I have created #5820 for futher discussion on this topic.  @vibhatha @whutjs @rui-mo can we continue to find a resolution there?